lib/Backends/OpenCL/OpenCLDeviceManager.h

```diff
@@ -0,0 +1,109 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef GLOW_BACKENDS_OPENCL_OPENCLDEVICEMANAGER_H
+#define GLOW_BACKENDS_OPENCL_OPENCLDEVICEMANAGER_H
+
+#include "glow/Backends/QueueBackedDeviceManager.h"
+
+#if defined(__APPLE__) || defined(__MACOSX)
+#include "OpenCL/opencl.h"
+#else
+#include <CL/cl.h>
+#endif
+
+namespace glow {
+namespace runtime {
+
+/// A class that contains an openCL device buffer. It frees the buffer when it
+/// is destroyed. Can be extended to store multiple buffers and rotate through
+/// them. Also tracks number of functions using this buffer.
+class OpenCLBuffer {
+  /// The OpenCL buffer being stored.
+  cl_mem buffer_;
+  /// Count of functions using this buffer.
+  unsigned int users_{0};
+  /// Size of the buffer in bytes.
+  const size_t size_{0};
+
+public:
+  ~OpenCLBuffer() { clReleaseMemObject(buffer_); }
+  OpenCLBuffer(cl_mem &buffer, size_t size) : buffer_(buffer), size_(size) {}
+  /// Returns the stored buffer.
+  cl_mem getBuffer() { return buffer_; }
+  /// Increment user count by 1 and return new count.
+  unsigned int incrementUsers() { return users_++; }
+  /// Decrement user count by 1 and return new count.
+  unsigned int decrementUsers() { return users_--; }
+  /// Get size of buffer in bytes.
+  size_t getSize() { return size_; }
+};
+
+/// A class controlling a single OpenCL device. Many OpenCLFunctions may be
+/// added, but only one inference is executed at a time.
+class OpenCLDeviceManager : public QueueBackedDeviceManager {
+  /// Compiled function list by name.
+  FunctionMapTy functions_;
+
+  /// Maximum available memory on the device.
+  uint64_t maxMemoryBytes_{0};
+
+  /// Amount of memory used by all models.
+  uint64_t usedMemoryBytes_{0};
```

Comment:
Should the operations on `usedMemoryBytes_` be atomic to avoid any possible race conditions?